src/main/java/server/Server.java

```diff
@@ -0,0 +1,102 @@
+package server;/*
+ * This Java source file was generated by the Gradle 'init' task.
+ */
+
+import Response.Response;
+
+import java.io.*;
+import java.net.*;
+import java.util.ArrayList;
+
+import static com.google.common.io.ByteStreams.toByteArray;
+
+public class Server {
+  private PrintWriter out;
+  private BufferedReader in;
+  private ServerSocket serverSocket;
+  private int PORT = 3000;
+  private Socket clientSocket;
+
+  public void go() throws IOException {
+    serverSocket = createSocket();
+
+    while (true) {
+      clientSocket = clientSocket(serverSocket);
+
+      out = createWriter(clientSocket);
+      in = createReader(clientSocket);
+      handleRequest(in, out);
+      closeConnection(in, out);
+      clientSocket.close();
+      System.out.println("Closing connection with testClient: " + clientSocket.getInetAddress());
+    }
+  }
+
+  public void handleRequest(BufferedReader in, PrintWriter out) {
+    try {
+      Response response = new Response();
+      String line;
+      ArrayList<String> responseText = new ArrayList();
+      while ((line = in.readLine()) != null) {
```

Comment:
I see this is a `BufferedStreamReader`.  [Check out these docs](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html).

When you use `readLine()`, you're imagining the request body as a series of lines of _text_.  _Usually_ that's the case, but it makes some assumptions that it will _always_ be the case.  

Check out these two methods, as well:
* [BufferedStreamReader#ready](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#ready--)
* [BufferedStreamReader#read](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedReader.html#read--)

The `ready` method tells you whether or not there's anything left to read.  In other words, it does the `!= null` check for you.

And `read` gives you the next character (as an `int` value, so you can coerce it yourself) of the stream.  Basically, what this method is doing, is letting you have access to the buffer of input without making assumptions about its shape.

Take this method for example:

```java
public String readRequestFromSocket(Socket clientSocket) {
  StringBuilder requestBuilder = new StringBuilder();
  BufferedStreamReader buffer = new BufferedStreamReader(
    new InputStreamReader(clientSocket.getInputStream(), StandardCharsets.UTF_8)
  );

  while (buffer.ready() || requestBuilder.length() == 0) {
    requestBuilder.append((char) buffer.read());
  }

  return requestBuilder.toString();
}
```

This does the following:
* Forces an initial attempt to read from a buffer if we haven't yet started our string builder (`requestBuilder.length() == 0`).
* Extract a string from a buffer, character by character (making no assumptions about request body shape).
* Coerces `int` stream into `char` values.  If we wanted to make a different marshaling system here, we could.  Meanwhile, we're also reading those `int` values out as `UTF_8`, which is a handy way to reduce encoding problems.

But even if you don't do it _exactly_ like this, it also does one other important thing: it splits your handling of input reading apart from output reading.

Basically, the writing and the reading are living together in the same loop here, and that's definitely going to cause problems.  Splitting them apart is probably a good first move for you:

1. First, get all of your request.
2. Interpret the request - my favorite thing to do here is to create some of my own code which reads the string content of the request, and then turns what I learned into plain Java objects.
3. Use the interpreted request to create a response.
4. Write the response back to the socket.